ast.go

```diff
@@ -1293,7 +1294,55 @@ func (node *Insert) String() string {
 	for _, row := range node.Rows {
 		rows = append(rows, row.String())
 	}
-	return fmt.Sprintf("insert into %s%s values %s%s", node.Table.String(), node.Columns.String(), strings.Join(rows, ", "), returning)
+	return fmt.Sprintf("insert into %s%s values %s%s%s", node.Table.String(), node.Columns.String(), strings.Join(rows, ", "), node.Upsert.String(), returning)
+}
+
+// Upsert represents an upsert clause, which is a list of on conflict clause.
+type Upsert []*OnConflictClause
+
+func (node Upsert) String() string {
+	if len(node) == 0 {
+		return ""
+	}
+
+	var clauses []string
+	for _, clause := range node {
+		clauses = append(clauses, fmt.Sprintf("on conflict%s", clause.String()))
+	}
+
+	return fmt.Sprintf(" %s", strings.Join(clauses, " "))
+}
+
+type OnConflictClause struct {
+	Target   *OnConflictTarget
+	DoUpdate *OnConflictUpdate
+}
+
+func (node *OnConflictClause) String() string {
+	var target string
+	if node.Target != nil {
+		target = fmt.Sprintf("%s%s", node.Target.Columns.String(), node.Target.Where.String())
+	}
+
+	if node.DoUpdate == nil {
+		return fmt.Sprintf("%s do nothing", target)
+	}
+
+	var exprs []string
+	for _, expr := range node.DoUpdate.Exprs {
+		exprs = append(exprs, fmt.Sprintf("%s = %s", expr.Column.String(), expr.Expr.String()))
+	}
+	return fmt.Sprintf("%s do update set %s%s", target, strings.Join(exprs, ", "), node.DoUpdate.Where.String())
+}
```

Comment:
I'm mostly using lowercase. For some reason, the `CREATE TABLE` end up with uppercase :( not sure why

I agree on keeping things consistent. I'll see how much work it is to change that

Reply:
Actually, I do think this is kind of an important thing to specify sooner rather than later. This is a bit of a discussion about canonicalization. I'd say we should pick one, and then actually specify it as our string representation. I don't have a strong opinion on _which_ to pick (lowercase looks nicer imo, but UPPERCASE is more standard), but we should pick one and stick with it. That way hashing etc is consistent.